diff -wru a/notlibc.c b/notlibc.c
--- a/notlibc.c	2016-09-25 15:21:10.000000000 -0500
+++ b/notlibc.c	2017-11-06 18:29:55.000000000 -0600
@@ -94,22 +94,6 @@
 //void matherr( void *exp ){report("matherr ignored\n");}
 
 
-int _read (int file, char *ptr, int len) { return read(file, ptr, len); }
-int _lseek (int file, int ptr, int dir) { return lseek(file, ptr, dir); }
-int _write ( int file, char *ptr, int len) {
-  int n=len;
-  if(file!=1 && file!=2) return -1;
-#ifndef NOSERIAL
-  while(n-->0) serial_putc(*ptr++);
-  serial_flush();
-#endif
-  return len;
-}
-int _close (int file) { return close(file); }
-caddr_t _sbrk (int incr) { return sbrk(incr); }
-int _open (const char *path, int flags, ...) { return open(path, flags, 0); }
-int _fstat (int file, struct stat *st) { st->st_mode = S_IFCHR; return 0; }
-int _isatty (int fd) { return fd>=0 && fd<=2; }
 int isatty (int fd) __attribute__((weak));
 int isatty (int fd) { return _isatty(fd); }
 int _fcntl(int fildes, int cmd, ...)
